Portlayer panic github.com/vmware/vic/lib/portlayer/network.(*Endpoint).copy on VC

The test container is hung on this command:
```
root         1  0.0  0.0   4344   636 ?        Ss   16:01   0:00 /bin/sh -c echo $CI_SCRIPT | base64 -d | /bin/sh -e
root         9  0.0  0.0   4344   688 ?        S    16:01   0:00 /bin/sh -e
root        51  0.0  0.0  11684  2500 ?        S    16:01   0:00 /bin/bash tests/integration-test.sh
root       182  0.2  1.0 142580 82132 ?        S    16:01   0:31 /usr/bin/python /usr/local/bin/pybot --removekeywords TAG:secret --exclude skip tests/test-cases
root      4563  0.0  0.0      0     0 ?        Zs   16:35   0:00 [sh] <defunct>
root     11039  0.0  0.0   4344   732 ?        S    17:15   0:00 sh -c docker -H 10.158.214.77:2376 --tls run --name verify-anon-1 jakedsouza/group-1-19-docker-verify-volume-files:1.0 ls /etc/example 2>&1
root     11040  0.0  0.1 200516 16340 ?        Sl   17:15   0:00 docker -H 10.158.214.77:2376 --tls run --name verify-anon-1 jakedsouza/group-1-19-docker-verify-volume-files:1.0 ls /etc/example
root     11047  0.0  0.0  20304  3308 pts/0    Ss   19:27   0:00 /bin/bash
root     11057  0.0  0.0  17508  2076 pts/0    R+   19:28   0:00 ps aux
```

The container id of concern:

`074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326`


Failure - personality:
```
Feb 22 2018 17:18:08.160Z INFO  container attach: stdout (074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326) done
Feb 22 2018 17:18:08.160Z INFO  container attach: stderr (074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326) done
Feb 22 2018 17:18:08.160Z INFO  cleaned up connections to 074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326. Checking errors
Feb 22 2018 17:18:08.160Z INFO  No error found. Returning nil...
Feb 22 2018 17:18:08.160Z DEBUG [ END ]  [vic/pkg/retry.DoWithConfig:73] [6.516497515s] 
Feb 22 2018 17:18:08.160Z DEBUG [ END ]  [vic/lib/apiservers/engine/backends.(*Container).ContainerAttach:1709] [6.516526515s] 074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326
Feb 22 2018 17:18:10.294Z DEBUG [ END ]  [vic/pkg/retry.DoWithConfig:73] [8.309730848s] 
Feb 22 2018 17:18:10.294Z DEBUG [ END ] op=296.26 [vic/lib/apiservers/engine/backends.(*Container).ContainerStart:906] [8.309765394s] 074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326
```

Pass - personality (example output from a passing run):
```
Feb 21 2018 22:48:10.314Z DEBUG Calling POST /v1.25/containers/f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d/start
Feb 21 2018 22:48:10.314Z DEBUG op=290.30: [NewOperation] op=290.30 [vic/lib/apiservers/engine/backends.(*Container).ContainerStart:905]
Feb 21 2018 22:48:10.314Z DEBUG [BEGIN] op=290.30 [vic/lib/apiservers/engine/backends.(*Container).ContainerStart:906] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:10.314Z DEBUG [BEGIN]  [vic/pkg/retry.DoWithConfig:73]
Feb 21 2018 22:48:10.314Z DEBUG op=290.30: Obtained container lock for f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:10.329Z DEBUG op=290.30: Binding network to container f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:10.329Z DEBUG [BEGIN]  [vic/lib/apiservers/engine/backends.(*Container).cleanupPortBindings:853] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:10.329Z DEBUG [ END ]  [vic/lib/apiservers/engine/backends.(*Container).cleanupPortBindings:853] [9.751µs] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:10.329Z DEBUG op=290.30: Setting container f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d state to running
Feb 21 2018 22:48:10.330Z DEBUG op=290.30: Commit container f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:16.534Z DEBUG [ END ]  [vic/pkg/retry.DoWithConfig:73] [6.220302911s] 
Feb 21 2018 22:48:16.534Z DEBUG [ END ] op=290.30 [vic/lib/apiservers/engine/backends.(*Container).ContainerStart:906] [6.220329195s] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:16.541Z DEBUG [BEGIN]  [vic/lib/apiservers/engine/backends.DockerEventPublisher.PublishEvent:220] Event Monitor received eventID(296.884) for container(f3e551c887af) - Started
Feb 21 2018 22:48:16.541Z DEBUG [ END ]  [vic/lib/apiservers/engine/backends.DockerEventPublisher.PublishEvent:220] [58.711µs] Event Monitor received eventID(296.884) for container(f3e551c887af) - Started
Feb 21 2018 22:48:16.562Z INFO  container attach: stderr (f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d) done
Feb 21 2018 22:48:16.562Z DEBUG * stdout attach end Wed Feb 21 22:48:16 UTC 2018
Feb 21 2018 22:48:16.562Z INFO  container attach: stdout (f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d) done
Feb 21 2018 22:48:16.562Z INFO  cleaned up connections to f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d. Checking errors
Feb 21 2018 22:48:16.562Z INFO  No error found. Returning nil...
Feb 21 2018 22:48:16.562Z DEBUG [ END ]  [vic/pkg/retry.DoWithConfig:73] [6.432662056s] 
Feb 21 2018 22:48:16.562Z DEBUG [ END ]  [vic/lib/apiservers/engine/backends.(*Container).ContainerAttach:1709] [6.432687101s] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:16.937Z DEBUG [BEGIN]  [vic/lib/apiservers/engine/backends.DockerEventPublisher.PublishEvent:220] Event Monitor received eventID(296.886) for container(f3e551c887af) - PoweredOff
Feb 21 2018 22:48:16.937Z DEBUG [ END ]  [vic/lib/apiservers/engine/backends.DockerEventPublisher.PublishEvent:220] [33.207µs] Event Monitor received eventID(296.886) for container(f3e551c887af) - PoweredOff
Feb 21 2018 22:48:16.937Z DEBUG [BEGIN]  [vic/lib/apiservers/engine/backends.(*ContainerProxy).exitCode:879]
Feb 21 2018 22:48:16.984Z DEBUG [ END ]  [vic/lib/apiservers/engine/backends.(*ContainerProxy).exitCode:879] [46.107997ms] 
Feb 21 2018 22:48:16.984Z INFO  Sending die event for container(f3e551c887af) with exitCode[0] - eventID(296.886)
Feb 21 2018 22:48:16.984Z DEBUG UnmapPorts for f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d: map[]
Feb 21 2018 22:48:16.988Z DEBUG Client context cancelled, stop sending events
```



Failure - portlayer:
```
Feb 22 2018 17:18:10.572Z DEBUG [ END ] op=442.473 [vic/lib/portlayer/network.(*Context).UnbindContainer:961] [71.199µs] UnbindContainer
Feb 22 2018 17:18:10.572Z DEBUG [BEGIN] op=442.473 [vic/lib/portlayer/exec.Commit:34] 074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326
Feb 22 2018 17:18:10.572Z INFO  op=442.473: Reconfigure: attempting update to 074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326 with change version "2018-02-22T17:18:02.11905Z" (poweredOff)
Feb 22 2018 17:18:10.789Z INFO  op=442.473: Reconfigure: committed update to 074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326 with change version: 2018-02-22T17:18:02.11905Z
Feb 22 2018 17:18:10.789Z INFO  op=442.473: Attempting to perform a guest reconfigure operation on (074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326)
Feb 22 2018 17:18:10.789Z DEBUG [BEGIN]  [vic/pkg/retry.DoWithConfig:73]
Feb 22 2018 17:18:10.789Z DEBUG [ END ]  [vic/pkg/retry.DoWithConfig:73] [6.198µs] 
Feb 22 2018 17:18:10.789Z DEBUG op=442.473: Update: updated change version from handle: 2018-02-22T17:18:02.11905Z
Feb 22 2018 17:18:10.789Z DEBUG [ END ] op=442.473 [vic/lib/portlayer/exec.Commit:34] [216.823085ms] 074b7ed656e57633397df644b879ef4ca317f79a1c907644d663d687854c0326
```

Pass - portlayer (example from a passing run):
```
Feb 21 2018 22:48:16.967Z DEBUG [ END ] op=296.887 [vic/lib/portlayer/network.(*Context).UnbindContainer:961] [61.62µs] UnbindContainer
Feb 21 2018 22:48:16.967Z DEBUG [BEGIN] op=296.887 [vic/lib/portlayer/exec.Commit:34] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:16.967Z INFO  op=296.887: Reconfigure: attempting update to f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d with change version "2018-02-21T22:48:11.203752Z" (poweredOff)
Feb 21 2018 22:48:16.968Z DEBUG [BEGIN] op=296.888 [vic/lib/portlayer/exec.(*containerBase).updates:119] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:16.968Z DEBUG op=296.888: Update: for f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d, refreshing from change version 2018-02-21T22:48:11.203752Z
Feb 21 2018 22:48:16.968Z DEBUG [BEGIN] op=296.888 [vic/pkg/vsphere/vm.(*VirtualMachine).Properties:519] VM(VirtualMachine:10532) Properties([config runtime])
Feb 21 2018 22:48:16.968Z DEBUG op=296.888: properties: [config runtime summary.runtime.connectionState]
Feb 21 2018 22:48:16.983Z DEBUG [ END ] op=296.888 [vic/pkg/vsphere/vm.(*VirtualMachine).Properties:519] [15.091843ms] VM(VirtualMachine:10532) Properties([config runtime])
Feb 21 2018 22:48:16.983Z DEBUG op=296.888: Update: for f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d, change version 2018-02-21T22:48:11.203752Z, extraconfig id: f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:16.983Z DEBUG [ END ] op=296.888 [vic/lib/portlayer/exec.(*containerBase).updates:119] [15.652628ms] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:16.983Z DEBUG [ END ] op=296.888 [vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).GetContainerInfoHandler:276] [19.360418ms] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
Feb 21 2018 22:48:17.105Z INFO  op=296.887: Reconfigure: committed update to f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d with change version: 2018-02-21T22:48:11.203752Z
Feb 21 2018 22:48:17.105Z INFO  op=296.887: Attempting to perform a guest reconfigure operation on (f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d)
Feb 21 2018 22:48:17.105Z DEBUG [BEGIN]  [vic/pkg/retry.DoWithConfig:73]
Feb 21 2018 22:48:17.105Z DEBUG [ END ]  [vic/pkg/retry.DoWithConfig:73] [6.108µs] 
Feb 21 2018 22:48:17.105Z DEBUG op=296.887: Update: updated change version from handle: 2018-02-21T22:48:11.203752Z
Feb 21 2018 22:48:17.105Z DEBUG [ END ] op=296.887 [vic/lib/portlayer/exec.Commit:34] [137.494467ms] f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d
```

From the looks of it, the portlayer output is generally the same, so it looks like the personality either missed a signal from the portlayer or got hung up somewhere else.

Mainly it looks like we missed this in the failing case:
```
Feb 21 2018 22:48:16.937Z DEBUG [BEGIN]  [vic/lib/apiservers/engine/backends.DockerEventPublisher.PublishEvent:220] Event Monitor received eventID(296.886) for container(f3e551c887af) - PoweredOff
Feb 21 2018 22:48:16.937Z DEBUG [ END ]  [vic/lib/apiservers/engine/backends.DockerEventPublisher.PublishEvent:220] [33.207µs] Event Monitor received eventID(296.886) for container(f3e551c887af) - PoweredOff
Feb 21 2018 22:48:16.937Z DEBUG [BEGIN]  [vic/lib/apiservers/engine/backends.(*ContainerProxy).exitCode:879]
Feb 21 2018 22:48:16.984Z DEBUG [ END ]  [vic/lib/apiservers/engine/backends.(*ContainerProxy).exitCode:879] [46.107997ms] 
Feb 21 2018 22:48:16.984Z INFO  Sending die event for container(f3e551c887af) with exitCode[0] - eventID(296.886)
Feb 21 2018 22:48:16.984Z DEBUG UnmapPorts for f3e551c887af4035e355b8900b9f805925f3fb0feb90992ffcf8838674a29e5d: map[]
Feb 21 2018 22:48:16.988Z DEBUG Client context cancelled, stop sending events
```
specifically -the exitCode, which presumably means that we never return an exitcode to the client...

[container-logs.zip](https://github.com/vmware/vic/files/1749435/container-logs.zip)

The portlayer should be publishing the stop event on it's event stream for the personality to pick up.
That looks like this (example is the created event):
```
Feb 23 2018 17:28:13.128Z DEBUG Found 2 subscribers to 298.39: events.ContainerEvent - Container 9c53755f93802108a5868c9ae845fd3930bc49b945d3c1690c9c76f8fc2ddc7f Created
Feb 23 2018 17:28:13.128Z DEBUG Event manager calling back to netCtx(0xc420ba08a0) for Event(298.39): events.ContainerEvent
Feb 23 2018 17:28:13.128Z DEBUG Event manager calling back to PLE-8546d0a8b0e9756010c4ce3607ed4608be7cfc69d559a851d579edafe60dbe27 for Event(298.39): events.ContainerEvent
```

However that PLE callback is not seen when the container in question shuts down and we can see that the event manager only has 1 subscriber at this time:
```
Feb 23 2018 17:44:56.979Z DEBUG Found 1 subscribers to 498.947: events.ContainerEvent - Container 9ba57c708ff6ea3e13f2365b9f30ca6a12166211c3341945ebb55111e6e55173 PoweredOff
Feb 23 2018 17:44:56.979Z DEBUG Event manager calling back to netCtx(0xc420b314a0) for Event(498.947): events.ContainerEvent
```

Looking for the last instance of a PLE event being processed we see this panic and then the portlayer restarts:
```
Feb 23 2018 17:34:25.844Z DEBUG Found 2 subscribers to 298.430: events.ContainerEvent - Container 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a PoweredOff
Feb 23 2018 17:34:25.844Z DEBUG Event manager calling back to netCtx(0xc420ba08a0) for Event(298.430): events.ContainerEvent
Feb 23 2018 17:34:25.844Z DEBUG Event manager calling back to PLE-8546d0a8b0e9756010c4ce3607ed4608be7cfc69d559a851d579edafe60dbe27 for Event(298.430): events.ContainerEvent
Feb 23 2018 17:34:25.850Z DEBUG op=298.433: [NewOperation] op=298.433 [vic/lib/portlayer/network.handleEvent:123]
Feb 23 2018 17:34:25.850Z INFO  op=298.433: Handling Event: 298.430
Feb 23 2018 17:34:25.859Z DEBUG [ END ] op=298.432 [vic/pkg/vsphere/vm.(*VirtualMachine).Properties:519] [15.250633ms] VM(VirtualMachine:vm-380) Properties([config runtime])
Feb 23 2018 17:34:25.859Z DEBUG op=298.432: Update: for 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a, change version 2018-02-23T17:34:12.409872Z, extraconfig id: 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:25.860Z DEBUG [ END ] op=298.432 [vic/lib/portlayer/exec.(*containerBase).updates:119] [16.01975ms] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:25.860Z DEBUG [ END ] op=298.432 [vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).GetContainerInfoHandler:276] [16.050236ms] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:25.861Z DEBUG op=298.434: [NewOperation] op=298.434 [vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).RemoveContainerHandler:242]
Feb 23 2018 17:34:25.861Z DEBUG [BEGIN] op=298.434 [vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).RemoveContainerHandler:243] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:25.861Z DEBUG [BEGIN] op=298.434 [vic/lib/portlayer/exec.(*Container).Remove:574] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:25.861Z DEBUG op=298.434: Updating container 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a state: Stopped->Removing
Feb 23 2018 17:34:25.861Z DEBUG [BEGIN] op=298.434 [vic/pkg/vsphere/vm.(*VirtualMachine).Properties:519] VM(VirtualMachine:vm-380) Properties([config.files.vmPathName])
Feb 23 2018 17:34:25.861Z DEBUG op=298.434: properties: [config.files.vmPathName summary.runtime.connectionState]
Feb 23 2018 17:34:25.861Z DEBUG op=298.435: [NewOperation] op=298.435 [vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).GetContainerInfoHandler:275]
Feb 23 2018 17:34:25.861Z DEBUG [BEGIN] op=298.435 [vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).GetContainerInfoHandler:276] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:25.881Z DEBUG [ END ] op=298.434 [vic/pkg/vsphere/vm.(*VirtualMachine).Properties:519] [19.869617ms] VM(VirtualMachine:vm-380) Properties([config.files.vmPathName])
Feb 23 2018 17:34:27.543Z DEBUG [ END ] op=298.434 [vic/lib/portlayer/exec.(*Container).Remove:574] [1.682278679s] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:27.548Z DEBUG [ END ] op=298.434 [vic/lib/apiservers/portlayer/restapi/handlers.(*ContainersHandlersImpl).RemoveContainerHandler:243] [1.686726403s] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:27.551Z DEBUG [BEGIN] op=298.433 [vic/lib/portlayer/exec.(*containerBase).updates:119] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:27.551Z DEBUG op=298.433: Update: for 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a, refreshing from change version 2018-02-23T17:34:12.409872Z
Feb 23 2018 17:34:27.551Z DEBUG [BEGIN] op=298.433 [vic/pkg/vsphere/vm.(*VirtualMachine).Properties:519] VM(VirtualMachine:vm-380) Properties([config runtime])
Feb 23 2018 17:34:27.551Z DEBUG op=298.433: properties: [config runtime summary.runtime.connectionState]
Feb 23 2018 17:34:27.551Z DEBUG Found 2 subscribers to 298.434: events.ContainerEvent - Container 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a Removed
Feb 23 2018 17:34:27.551Z DEBUG Event manager calling back to netCtx(0xc420ba08a0) for Event(298.434): events.ContainerEvent
Feb 23 2018 17:34:27.551Z DEBUG Event manager calling back to PLE-8546d0a8b0e9756010c4ce3607ed4608be7cfc69d559a851d579edafe60dbe27 for Event(298.434): events.ContainerEvent
Feb 23 2018 17:34:27.564Z DEBUG [ END ] op=298.433 [vic/pkg/vsphere/vm.(*VirtualMachine).Properties:519] [13.770825ms] VM(VirtualMachine:vm-380) Properties([config runtime])
Feb 23 2018 17:34:27.564Z DEBUG [ END ] op=298.433 [vic/lib/portlayer/exec.(*containerBase).updates:119] [13.816529ms] 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a
Feb 23 2018 17:34:27.564Z ERROR op=298.433: Update: unable to update container 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a: ServerFaultCode: The object 'vim.VirtualMachine:vm-380' has already been deleted or has not been completely created
Feb 23 2018 17:34:27.564Z ERROR op=298.433: refreshing container 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a failed: ServerFaultCode: The object 'vim.VirtualMachine:vm-380' has already been deleted or has not been completely created
Feb 23 2018 17:34:27.564Z DEBUG [BEGIN] op=298.433 [vic/lib/portlayer/network.(*Context).RemoveIDFromScopes:921]
Feb 23 2018 17:34:27.564Z DEBUG op=298.436: [NewOperation] op=298.436 [vic/lib/portlayer/network.(*Scope).RemoveContainer:219]
Feb 23 2018 17:34:27.564Z INFO  Released IP range [172.28.0.2, 172.28.0.2]
Feb 23 2018 17:34:27.565Z DEBUG op=298.436: Container 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a removed from the scope barnet(53e3d374ced5741befc62c5424f6a061e64f9dc27ef1d114c5a8e47870c94738)
Feb 23 2018 17:34:27.565Z DEBUG op=298.437: [NewOperation] op=298.437 [vic/lib/portlayer/network.(*Scope).RemoveContainer:219]
Feb 23 2018 17:34:27.565Z INFO  Released IP range [172.16.0.2, 172.16.0.2]
Feb 23 2018 17:34:27.565Z DEBUG op=298.437: Container 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a removed from the scope bridge(07722e164d845841f90a43ce6bddc5679e2a962921a0356b4c74c5773bb1910f)
Feb 23 2018 17:34:27.565Z DEBUG op=298.438: [NewOperation] op=298.438 [vic/lib/portlayer/network.(*Scope).RemoveContainer:219]
Feb 23 2018 17:34:27.565Z INFO  Released IP range [172.27.0.2, 172.27.0.2]
Feb 23 2018 17:34:27.565Z DEBUG op=298.438: Container 7a48efde8f4e271eddcca08d5ff9c02e7801b98f86dcb940328127cbf33b2e5a removed from the scope foonet(1bcc51df6ae59d4ca517543fbd1bfb17444e03094b581f7b38cd3db06a75da29)
Feb 23 2018 17:34:27.565Z DEBUG [ END ] op=298.433 [vic/lib/portlayer/network.(*Context).RemoveIDFromScopes:921] [191.348µs] 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd8615a]

goroutine 26 [running]:
github.com/vmware/vic/lib/portlayer/network.(*Endpoint).copy(0x0, 0xc4212611d0)
	/go/src/github.com/vmware/vic/lib/portlayer/network/endpoint.go:190 +0x4a
github.com/vmware/vic/lib/portlayer/network.(*Context).RemoveIDFromScopes(0xc420ba08a0, 0x1dd6080, 0xc42039a180, 0xc4202085f0, 0x1, 0x1, 0xc420fd2068, 0x7, 0x0, 0xc4212dad40, ...)
	/go/src/github.com/vmware/vic/lib/portlayer/network/context.go:943 +0x28b
github.com/vmware/vic/lib/portlayer/network.handleEvent(0xc420ba08a0, 0x1dd9ac0, 0xc42000e2a0)
	/go/src/github.com/vmware/vic/lib/portlayer/network/network.go:128 +0x8e8
github.com/vmware/vic/lib/portlayer/network.engageContext.func1(0x1dd9ac0, 0xc42000e2a0)
	/go/src/github.com/vmware/vic/lib/portlayer/network/network.go:176 +0x3e
github.com/vmware/vic/lib/portlayer/event.newSubscriber.func1(0xc420a00d20)
	/go/src/github.com/vmware/vic/lib/portlayer/event/subscriber.go:113 +0x194
created by github.com/vmware/vic/lib/portlayer/event.newSubscriber
	/go/src/github.com/vmware/vic/lib/portlayer/event/subscriber.go:119 +0x12a
time="2018-02-23T17:34:31Z" level=info msg="Launching portlayer server pprof server on 127.0.0.1:6063" 
time="2018-02-23T17:34:31Z" level=info msg="{Formatter:0xc4203bbc30 Level:debug Syslog:<nil>}" 
```
Items of interest in this fragment:
1. we have not processed the network unbind from the container Stopped event before receiving the Removed event
2. we see the container being removed from **three** scopes the the following addresses:
  `172.16.0.2`
  `172.28.0.2`
  `172.27.0.2`
3. only **one** scope bind operation was performed for that container and that was with address `172.16.0.3`.

Fragment for scopes bind:
```
Feb 23 2018 17:44:27.219Z DEBUG op=498.943: [NewOperation] op=498.943 [vic/lib/apiservers/portlayer/restapi/handlers.(*ScopesHandlersImpl).ScopesBindContainer:273]
Feb 23 2018 17:44:27.219Z DEBUG [BEGIN] op=498.943 [vic/lib/apiservers/portlayer/restapi/handlers.(*ScopesHandlersImpl).ScopesBindContainer:274] handle(2fbb2f642cd99334c023cee9ced43f03)
Feb 23 2018 17:44:27.219Z DEBUG [BEGIN] op=498.943 [vic/lib/portlayer/network.(*Context).BindContainer:661]
Feb 23 2018 17:44:27.219Z DEBUG [BEGIN]  [vic/lib/portlayer/network.(*Context).container:854]
Feb 23 2018 17:44:27.219Z DEBUG [ END ]  [vic/lib/portlayer/network.(*Context).container:854] [9.365µs] 
Feb 23 2018 17:44:27.219Z DEBUG op=498.943: BindContainer found container 9ba57c708ff6ea3e13f2365b9f30ca6a12166211c3341945ebb55111e6e55173 with dynamic endpoint and unassigned IP
Feb 23 2018 17:44:27.219Z DEBUG op=498.944: [NewOperation] op=498.944 [vic/lib/portlayer/network.(*Scope).AddContainer:188]
Feb 23 2018 17:44:27.219Z DEBUG op=498.944: Adding container 9ba57c708ff6ea3e13f2365b9f30ca6a12166211c3341945ebb55111e6e55173 to the scope bridge(15ed968df84f9879e4bd3d16a359f37b04358f524df6591c4fed69b131ac1022)
Feb 23 2018 17:44:27.219Z INFO  Reserving IP range [172.16.0.3, 172.16.0.3]
Feb 23 2018 17:44:27.219Z DEBUG adding alias bridge:anjunabeats -> anjunabeats
Feb 23 2018 17:44:27.219Z DEBUG adding alias bridge:9ba57c708ff6 -> anjunabeats
Feb 23 2018 17:44:27.219Z DEBUG [ END ] op=498.943 [vic/lib/portlayer/network.(*Context).BindContainer:661] [109.903µs] 
Feb 23 2018 17:44:27.219Z DEBUG [ END ] op=498.943 [vic/lib/apiservers/portlayer/restapi/handlers.(*ScopesHandlersImpl).ScopesBindContainer:274] [129.112µs] handle(2fbb2f642cd99334c023cee9ced43f03)
```

We can see that after the portlayer restart we unbind only the one correct address associated with the container:
```
Feb 23 2018 17:44:57.004Z DEBUG [BEGIN] op=498.948 [vic/lib/portlayer/network.(*Context).UnbindContainer:961] UnbindContainer
Feb 23 2018 17:44:57.004Z DEBUG op=498.948: Trying to unbind container: 9ba57c708ff6ea3e13f2365b9f30ca6a12166211c3341945ebb55111e6e55173
Feb 23 2018 17:44:57.004Z DEBUG [BEGIN]  [vic/lib/portlayer/network.(*Context).container:854]
Feb 23 2018 17:44:57.004Z DEBUG [ END ]  [vic/lib/portlayer/network.(*Context).container:854] [7.857µs] 
Feb 23 2018 17:44:57.004Z DEBUG op=498.948: Removing endpoints from container 9ba57c708ff6ea3e13f2365b9f30ca6a12166211c3341945ebb55111e6e55173
Feb 23 2018 17:44:57.004Z DEBUG op=498.949: [NewOperation] op=498.949 [vic/lib/portlayer/network.(*Scope).RemoveContainer:219]
Feb 23 2018 17:44:57.004Z INFO  Released IP range [172.16.0.3, 172.16.0.3]
Feb 23 2018 17:44:57.004Z DEBUG op=498.949: Container 9ba57c708ff6ea3e13f2365b9f30ca6a12166211c3341945ebb55111e6e55173 removed from the scope bridge(15ed968df84f9879e4bd3d16a359f37b04358f524df6591c4fed69b131ac1022)
Feb 23 2018 17:44:57.004Z DEBUG [ END ] op=498.948 [vic/lib/portlayer/network.(*Context).UnbindContainer:961] [74.845µs] UnbindContainer
```

As such there are two primary issues here:
- [ ] why did the network handleEvent (`opid=298.433`) go off the deep end?
- [ ] why didn't the portlayer event stream re-register for events after the restart

@mhagen-vmware I recommend adding checks for panics in the logs as part of the general test validation - if the only reason we're seeing this panic is because of a distant secondary effect in a different component then it's possible we are missing others.

From https://ci-vic.vmware.com/vmware/vic/17145/8
Completely different test, looks like same stacktrace:
```
Mar  1 2018 18:39:12.226Z DEBUG op=301.438: Container ebd6b691b53420716d7486e842f96f3883ddb15513aeea30aed97c1c3c3f9f1e removed from the scope foonet(44b4c05b20371603400cb14b9f97a17b22268818c5b5596f41faf1da0b724f42)
Mar  1 2018 18:39:12.226Z DEBUG [ END ] op=301.431 [vic/lib/portlayer/network.(*Context).RemoveIDFromScopes:921] [119.701µs] 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xd8615a]

goroutine 25 [running]:
github.com/vmware/vic/lib/portlayer/network.(*Endpoint).copy(0x0, 0xc4208831d0)
	/go/src/github.com/vmware/vic/lib/portlayer/network/endpoint.go:190 +0x4a
github.com/vmware/vic/lib/portlayer/network.(*Context).RemoveIDFromScopes(0xc420ae0cc0, 0x1dd6080, 0xc420cf48d0, 0xc420836e10, 0x1, 0x1, 0xc420e76df8, 0x7, 0x0, 0xc4215c8180, ...)
	/go/src/github.com/vmware/vic/lib/portlayer/network/context.go:943 +0x28b
github.com/vmware/vic/lib/portlayer/network.handleEvent(0xc420ae0cc0, 0x1dd9ac0, 0xc4204452e0)
	/go/src/github.com/vmware/vic/lib/portlayer/network/network.go:128 +0x8e8
github.com/vmware/vic/lib/portlayer/network.engageContext.func1(0x1dd9ac0, 0xc4204452e0)
	/go/src/github.com/vmware/vic/lib/portlayer/network/network.go:176 +0x3e
github.com/vmware/vic/lib/portlayer/event.newSubscriber.func1(0xc420360870)
	/go/src/github.com/vmware/vic/lib/portlayer/event/subscriber.go:113 +0x194
created by github.com/vmware/vic/lib/portlayer/event.newSubscriber
	/go/src/github.com/vmware/vic/lib/portlayer/event/subscriber.go:119 +0x12a
time="2018-03-01T18:39:16Z" level=info msg="Launching portlayer server pprof server on 127.0.0.1:6063" 
```

--- a/lib/portlayer/network/context.go
+++ b/lib/portlayer/network/context.go
@@ -933,7 +933,7 @@ func (c *Context) RemoveIDFromScopes(op trace.Operation, id string) ([]*Endpoint
 		return nil, nil // not bound
 	}
 
-	// aliases to remove
+	// marshal the work so we're not iterating over a mutating structure
 	var aliases []string
 	var endpoints []*Endpoint
 	for _, ne := range con.endpoints {
@@ -942,12 +942,15 @@ func (c *Context) RemoveIDFromScopes(op trace.Operation, id string) ([]*Endpoint
 		// save the endpoint info
 		e := con.Endpoint(s).copy()
 
-		if err := s.RemoveContainer(con); err != nil {
-			return nil, err
-		}
-
 		aliases = append(aliases, c.populateAliases(con, s, e)...)
 		endpoints = append(endpoints, e)
+	}
+
+	// remove the container from all bound scopes
+	for _, ne := range endpoints {
+		if err := ne.scope.RemoveContainer(con); err != nil {
+			return nil, err
+		}
 	}
 
 	c.removeAliases(aliases, con)

--- a/lib/portlayer/network/context.go
+++ b/lib/portlayer/network/context.go
@@ -948,6 +948,7 @@ func (c *Context) RemoveIDFromScopes(op trace.Operation, id string) ([]*Endpoint
 
 	// remove the container from all bound scopes
 	for _, ne := range endpoints {
+		// this modifies the con.endpoint list so iteration has to occur off the copy
 		if err := ne.scope.RemoveContainer(con); err != nil {
 			return nil, err
 		}

